provisioning/device/src/transport_state_machine.ts

```diff
@@ -0,0 +1,282 @@
+          disconnect: (callback) => {
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_025: [ If `disconnect` is called while disconnected, it shall immediately call its `callback`. ] */
+            // nothing to do.
+            callback();
+          }
+        },
+        connecting: {
+          _onEnter: (callback) => {
+            debug('entering connecting state');
+            debug('calling transport _doConnectForFsm');
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_002: [ `connect` shall call `_doConnectForFsm`. ] */
+            this._doConnectForFsm((err) => {
+              /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_003: [ If `_doConnectForFsm` fails, then `connect` shall fail. ] */
+              if (err) {
+                /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_011: [ `register` shall fail if the connection fails. ] */
+                this._fsm.transition('disconnecting', callback, err);
+              } else {
+                this._fsm.transition('connected', callback);
+              }
+            });
+          },
+          '*': () => this._fsm.deferUntilTransition()
+        },
+        connected: {
+          _onEnter: (callback, err, body, result) => {
+            debug('entering connected state');
+            callback(err, body, result);
+          },
+          connect: (callback) => {
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_004: [ If the transport is already connected, then `connect` shall do nothing and call the `callback` immediately. ] */
+            // nothing to do.
+            callback();
+          },
+          disconnect: (callback) => {
+            this._fsm.transition('disconnecting', callback);
+          },
+          register: (callback, registrationId, authorization, requestBody, forceRegistration) => {
+            this._fsm.transition('sendingRegistrationRequest', callback, registrationId, authorization, requestBody, forceRegistration);
+          },
+        },
+        sendingRegistrationReqest: {
+          _onEnter: (callback, registrationId, authorization, requestBody, forceRegistration) => {
+            debug('entering sendingRegistrationRequest state');
+            this._registrationCallback = callback;
+            debug('calling transport _doFirstRegistrationRequestForFsm');
+            /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_012: [ `register` shall call `_doFirstRegistrationRequestForFsm`. ] */
+            this._doFirstRegistrationRequestForFsm(registrationId, authorization, requestBody, forceRegistration, (err, responseBody, result, pollingInterval) => {
+              if (this._registrationInProgress()) { // make sure we weren't cancelled before doing something with the response
+                if (err) {
+                  /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_013: [ If the registration request fails, `register` shall fail. ] */
```

Comment:
> `register` shall fail [](start = 117, length = 22)

How does this percolate out?

Reply:
These state machine have a strange way of returning failure.  The state you're in doesn't always match the failure you return.  For example, this transition says "move to the connected state, but also pass the error to it.  (ignore the fact that we might not want to stay connected for the moment).  Since the connected _onEnter function accepts a callback and an err, we can tacitly assume that the connect state will do something with this error.  In our case, it calls the callback because it's at the end of a chain of state change.  In other cases, it might pass it onto another state which will call the callback..

All these handlers are not type checked by TS at all.  I'm already getting spoiled because I miss that and it's not always easy to get the types correct.  

I'm going to make the return path more explicit by adding additional states for success and failure.

---
In reply to: [146401709](https://github.com/Azure/azure-iot-sdk-node/pull/115#discussion_r146401709) [](ancestors = 146401709)

Reply:
you can get help from the TS compiler by adding types in the signatures if you want.

---
In reply to: [147129572](https://github.com/Azure/azure-iot-sdk-node/pull/115#discussion_r147129572) [](ancestors = 147129572,146401709)